exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,407 @@
+				errList = append(errList, err)
+			}
+			unstructuredObj := &unstructured.Unstructured{Object: raw}
+
+			err = r.patchOwnerRefToResource(ctx, clusterResourceSet, unstructuredObj)
+			if err != nil {
+				logger.Error(err, "Failed to patch ClusterResourceSet as resource owner reference",
+					"Resource type", resourceConfigMap.Kind, "Resource name", resourceConfigMap.Name)
+				errList = append(errList, err)
+			}
+
+			// Since maps are not ordered, we need to order them to get the same hash at each reconcile.
+			keys := make([]string, 0)
+			for key := range resourceConfigMap.Data {
+				keys = append(keys, key)
+			}
+			sort.Strings(keys)
+
+			for _, key := range keys {
+				dataList = append(dataList, []byte(resourceConfigMap.Data[key]))
+			}
+
+		case string(addonsv1.SecretClusterResourceSetResourceKind):
+			// Set status in ClusterResourceSetBinding in case of early continue due to a failure. Do this for only supported kinds.
+			bindings.Resources[resourceKindName] = generateResourceStatus(false, "")
+
+			resourceSecret, err := getSecret(context.Background(), r.Client, resourceName)
+			if err != nil {
+				logger.Error(err, "Failed to find Secret resource", "Secret name", resourceName)
+				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.MissingResourceReason, clusterv1.ConditionSeverityWarning, err.Error())
+				errList = append(errList, err)
+				continue
+			}
+
+			if resourceSecret.Type != addonsv1.ClusterResourceSetSecretType {
+				logger.Error(err, "Unsupported ClusterResourceSet Secret type", "Resource type", resourceSecret.Type)
+				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.WrongSecretTypeReason, clusterv1.ConditionSeverityWarning, "")
+				errList = append(errList, err)
+				continue
+			}
+
+			// Convert the object to unstructured to compare against our before copy.
+			raw, err := runtime.DefaultUnstructuredConverter.ToUnstructured(resourceSecret.DeepCopyObject())
+			if err != nil {
+				logger.Error(err, "Failed to set ClusterResourceSet as resource owner reference",
+					"Resource type", resourceSecret.Type, "Resource name", resourceSecret.Name)
+				errList = append(errList, err)
+			}
+			unstructuredObj := &unstructured.Unstructured{Object: raw}
+
+			err = r.patchOwnerRefToResource(ctx, clusterResourceSet, unstructuredObj)
+			if err != nil {
+				logger.Error(err, "Failed to patch ClusterResourceSet as resource owner reference",
+					"Resource type", resourceSecret.Type, "Resource name", resourceSecret.Name)
+				errList = append(errList, err)
+			}
+			// Since maps are not ordered, we need to order them to get the same hash at each reconcile.
+			keys := make([]string, 0)
+			for key := range resourceSecret.Data {
+				keys = append(keys, key)
+			}
+			sort.Strings(keys)
+
+			for _, key := range keys {
+				dataList = append(dataList, resourceSecret.Data[key])
+			}
```

Comment:
Deduplicate this with the above by moving this logic outside the switch/case. Even better, move the switch to an internal method.